File: haskell/app/Game/Input.hs
module Game.Input where

import Data.Map.Strict

import Input

data Move
  = MoveLeft
  | MoveRight
  | MoveDown
  deriving Show

mkMove :: Input -> Maybe Move
mkMove input
  | isPressed $ keyboard input ! KbDown = Just MoveDown
  | otherwise =
    let
      moveLeft  = justPressed $ keyboard input ! KbLeft
      moveRight = justPressed $ keyboard input ! KbRight
    in
      case (moveLeft, moveRight) of
        (True , False) -> Just MoveLeft
        (False, True ) -> Just MoveRight
        _              -> Nothing

data Turn
  = TurnLeft
  | TurnRight
  deriving Show

mkTurn :: Input -> Maybe Turn
mkTurn input =
  let
    turnLeft  = justPressed $ mouse input ! MouseLeft
    turnRight = justPressed $ mouse input ! MouseRight
  in
    case (turnLeft, turnRight) of
      (True , False) -> Just TurnLeft
      (False, True ) -> Just TurnRight
      _              -> Nothing

File: rust/src/game/input.rs
use lume::Input;

#[derive(Copy, Clone, PartialEq, Eq)]
pub enum Move {
    Left,
    Right,
    Down,
}

impl TryFrom<&Input> for Move {
    type Error = ();

    fn try_from(input: &Input) -> Result<Self, Self::Error> {
        let move_down = input.keyboard.down.is_pressed();
        if move_down {
            Ok(Move::Down)
        } else {
            let move_left = input.keyboard.left.just_pressed();
            let move_right = input.keyboard.right.just_pressed();
            match (move_left, move_right) {
                (true, false) => Ok(Move::Left),
                (false, true) => Ok(Move::Right),
                _             => Err(()),
            }
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Turn {
    Left,
    Right,
}

impl TryFrom<&Input> for Turn {
    type Error = ();

    fn try_from(input: &Input) -> Result<Self, Self::Error> {
        let left = input.mouse.left.just_pressed();
        let right = input.mouse.right.just_pressed();
        match (left, right) {
            (true, false) => Ok(Turn::Left),
            (false, true) => Ok(Turn::Right),
            _             => Err(()),
        }
    }
}